brain_observatory_analysis/ophys/data_formatting.py

```diff
@@ -0,0 +1,533 @@
+        temp_df['bisect_layer'] = lamf_group.expt_table.loc[oeid].bisect_layer
+        temp_df.set_index('cell_specimen_id', inplace=True)
+        if run_pre_gray:
+            if trace_type=='dff':
+                temp_df['trace'] = trace_df.dff.apply(lambda x: x[:min_start_ind]).values
+            else:  # Other types already dealt with from above
+                temp_df['trace'] = trace_df.events.apply(lambda x: x[:min_start_ind]).values
+            temp_df['timepoints'] = [timestamps[:min_start_ind]] * len(csid)
+            sorted_temp_df = temp_df.sort_index(inplace=False)
+            trace_df_gray_pre_task = pd.concat([trace_df_gray_pre_task, sorted_temp_df])
+        if run_post_gray:
+            if trace_type=='dff':
+                temp_df['trace'] = trace_df.dff.apply(lambda x: x[max_end_ind:min_post_gray_end_ind]).values
+            else:
+                temp_df['trace'] = trace_df.events.apply(lambda x: x[max_end_ind:min_post_gray_end_ind]).values
+            temp_df['timepoints'] = [timestamps[max_end_ind:min_post_gray_end_ind]] * len(csid)
+            sorted_temp_df = temp_df.sort_index(inplace=False)
+            trace_df_gray_post_task = pd.concat([trace_df_gray_post_task, sorted_temp_df])
+            if run_fingerprint:
+                if trace_type=='dff':
+                    temp_df['trace'] = trace_df.dff.apply(lambda x: x[min_post_gray_end_ind:]).values
+                else:
+                    temp_df['trace'] = trace_df.events.apply(lambda x: x[min_post_gray_end_ind:]).values
+                temp_df['timepoints'] = [timestamps[min_post_gray_end_ind:]] * len(csid)
+                sorted_temp_df = temp_df.sort_index(inplace=False)
+                trace_df_fingerprint = pd.concat([trace_df_fingerprint, sorted_temp_df])
+    return trace_df_gray_pre_task, trace_df_gray_post_task, trace_df_fingerprint
+
+
+# A function to get non-auto-rewarded start and end times
+def get_non_auto_rewarded_start_end_times(stim_df):
+    """ Get start and end times of non-auto-rewarded stimuli
+
+    Parameters
+    ----------
+    stim_df: pandas.DataFrame
+        A dataframe containing stimulus information
+    
+    Returns
+    -------
+    start_times: list
+        A list of start times of non-auto-rewarded stimuli
+    stop_times: list
+        A list of stop times of non-auto-rewarded stimuli
+    """
+    auto_rewarded_inds = np.where(stim_df.auto_rewarded == True)[0]  # noqa: E712
+    diff_ar_inds = np.diff(auto_rewarded_inds)
+    skip_ar_inds = np.where(diff_ar_inds > 1)[0]
+    start_times = []  # inclusive
+    stop_times = []  # inclusive
+    for skip_ar_ind in skip_ar_inds:
+        start_times.append(stim_df.iloc[auto_rewarded_inds[skip_ar_ind] + 1].start_time)  # removing those before the first auto-rewarded stim
+        stop_times.append(stim_df.iloc[auto_rewarded_inds[skip_ar_ind + 1]].stop_time)
+    if auto_rewarded_inds[-1] != len(stim_df) - 1:
```

Comment:
when is this condition used? and why? (maybe add comment, why this if statement is here)

Reply:
This is to treat when the last stimulus is not auto-rewarded. In this case, we need to add the segments from one after the last auto-rewarded stimulus till the last stimulus. I left the comments as well.